### PR TITLE
Add std.mem.zeroes to the standard library

### PR DESCRIPTION
**Pull request by FireFox317**
I'm not sure if we should only allow extern structs or specific types, but this is a start.

See: https://github.com/ziglang/zig/issues/2257#issuecomment-570748449

### CODE REVIEW

**Comment by JesseRMeyer** [comment]
Is returning stack memory of Item undefined behavior in Zig once zero's scope returns?

**Comment by andrewrk** [comment]
> Is returning stack memory of Item undefined behavior in Zig once zero's scope returns?

It's a by-value return, so it's fine. The memory is provided by the caller and the data is copied into it. What you have to look out for is taking the address of stack memory, e.g. with `&`.

**Comment by JesseRMeyer** [comment]
> > Is returning stack memory of Item undefined behavior in Zig once zero's scope returns?
> 
> It's a by-value return, so it's fine. The memory is provided by the caller and the data is copied into it. What you have to look out for is taking the address of stack memory, e.g. with `&`.

Yeah, I _just_ noticed that.  Thought I first saw an address return but my eyes must be fuzzy today.

**Comment by LemonBoy** [comment]
Why is it limited to `extern struct`?

**Comment by FireFox317** [comment]
@LemonBoy because I'm not sure if normal/packed structs can also be zeroed like this. Because you can get null pointers and stuff, correct me if I'm wrong

**Comment by LemonBoy** [comment]
Indeed, I had something less safe in mind such as a member + return like the `secureZero` function declared below.

**Comment by andrewrk** [comment]
I do think this function should be discouraged for non-extern, non-packed structs (even if a reflection-based version exists that writes 0 memberwise).

I intentionally removed the `zeroes` keyword from the language because it allows us to put extra fields into structs for various reasons. Safety, hot code swapping, incremental compilation, and profiling, to name a few.

Writing zeroes to a non-packed, non-extern struct would clobber the safety field that allows safety of `@ptrCast` to work, for example. And it would clobber the detection of whether the struct was `undefined`.

File: lib/std/mem.zig [hunk 1]
```diff
@@ -273,6 +273,33 @@ pub fn set(comptime T: type, dest: []T, value: T) void {
...
+/// This can be used to zero initialize a C-struct.
+pub fn zeroes(comptime T: type) T {
+    if (@sizeOf(T) == 0) return T{};
+
+    if (comptime meta.containerLayout(T) != .Extern) {
+        @compileError("TODO: Currently this only works for extern types");
+    }
+
+    var item: T = undefined;
+    @memset(@ptrCast([*]u8, &item), 0, @sizeOf(T));
```
Comment (daurnimator) [inline lib/std/mem.zig]: Is this defined? e.g. if `T` was a `u24` then the size is 3 but `@sizeOf` returns 4.
See e.g. #4093

File: lib/std/mem.zig [hunk 2]
```diff
@@ -273,6 +273,33 @@ pub fn set(comptime T: type, dest: []T, value: T) void {
         d.* = value;
 }
 
+/// Zero initializes the type.
+/// This can be used to zero initialize a C-struct.
+pub fn zeroes(comptime T: type) T {
+    if (@sizeOf(T) == 0) return T{};
+
+    if (comptime meta.containerLayout(T) != .Extern) {
```
Comment (daurnimator) [inline lib/std/mem.zig]: Should allow packed as well?

Comment (andrewrk) [reply lib/std/mem.zig]: not necessary for merge, made complicated by #3133

File: lib/std/mem.zig (hunk 1 above)
Comment (andrewrk) [reply lib/std/mem.zig]: The code is correct. `@sizeOf` gives the number of bytes the type takes up in memory. This function is valid for types which have well-defined memory layout, which includes extern structs (the main use case).

**Comment by FireFox317** [comment]
Feel free to ask me to add anything. I can also allow packed structs and add a note regarding #3133 if that would be desired.

**Comment by andrewrk** [comment]
> Feel free to ask me to add anything. I can also allow packed structs and add a note regarding #3133 if that would be desired.

This is a useful first implementation. Further PRs are welcome to improve it as well :)

**Pull request merged**